requests-for-comments/2024-07-01-alerting.md

```diff
@@ -0,0 +1,160 @@
+trigger alerts based on their data. The system will evaluate these conditions at regular intervals and notify users
+when conditions are met. The design includes detailed logging for debugging and auditing, and efficient handling of
+task execution using Celery.
+
+#### Goals
+
+1. **Condition Evaluation**: Define and evaluate alert conditions at specified intervals.
+2. **Notification System**: Notify users when alert conditions are met.
+3. **Logging and Auditing**: Store detailed logs of each alert check.
+4. **Efficiency and Scalability**: Handle tasks efficiently to reduce load on the system.
+
+#### Architecture
+
+1. **Models**:
+    - **Alert**: Represents the alert configuration.
+    - **AlertCheck**: Logs each evaluation of an alert, including results and notification statuses.
+
+2. **Task Execution**:
+    - Use Celery to handle periodic checks and notifications.
+    - Implement concurrency limits and task chaining for efficient processing.
+
+#### Models
+
+##### Alert Model
+
+```python
+class Alert(models.Model):
+    team = models.ForeignKey("Team", on_delete=models.CASCADE)
+    insight = models.ForeignKey("posthog.Insight", on_delete=models.CASCADE)
+    name = models.CharField(max_length=100)
+    target_value = models.TextField()
+    anomaly_condition = models.JSONField(default=dict)
+```
+
+Credits: @nikitaevg in https://github.com/PostHog/posthog/pull/22554
+
+This main model is responsible for storing alert configurations. We may consider adding fields for alert frequency,
+notification preferences, and other settings based on user requirements. Future integration with CDP is also tbd.
+
+##### AlertCheck Model
+
+The `AlertCheck` model logs each evaluation of an alert:
+- `alert`: Foreign key to the Alert model.
+- `created`: Time of the check.
+- `calculated_value`: Result of the check.
+- `anomaly_condition`: Copy of the condition for the alert in case original got updated.
+- `threshold_met`: Whether the condition was met.
+- `notification_sent`: Whether notification was sent.
+- `error_message`: Any errors encountered.
+
+This model provides detailed logging for debugging and auditing purposes. It gives accountability and transparency to 
+users by recording the results of each alert evaluation.
+Users may ask why an alert was triggered and if the system is working as expected. The `AlertCheck` model helps answer 
+these questions by providing a detailed history of alert evaluations.
+
+We might consider discarding old `AlertCheck` records after a certain period to manage database size.
+
+#### Task Execution with Celery
+
+##### Why Use Groups, Chains, and Chunks
+
+**Groups**: Allow multiple tasks to be executed in parallel, providing the ability to run concurrent checks for 
+different alerts. By creating groups of tasks, we can limit the number of parallel operations to control system load 
+and ensure efficient resource utilization.
```

Comment:
I wonder maybe [limit_concurrency](https://github.com/PostHog/posthog/blame/4c4a2733987fda5619f648e3aafbc679f4005ea4/posthog/clickhouse/client/limit.py#L47) (that you introduced) would work here? Basically that's all we need - limit the number of tasks in flight, no?

If we add limit_concurrency and simply process all alerts in a `for`, IMO it would have the same effect as if we used chains + groups + chunks, but it would be simpler, or do I miss something?

Reply:
> the trick here (however we do it) is to limit the number of running tasks such that if the number of celery pods scale the max number of tasks does not.

IIUC the limit_concurrency implementation, it looks at the number of tasks in flight and drops the new one if it's above the limit, so I'd say it will keep the limit if we increase the number of pods, right?

> maybe teaching everyone to suck eggs 🙈

I haven't worked with Celery before, so not at all! What an odd expression though.

Reply:
> What an odd expression though. 

🤣 

i have a bad habit of speaking in idioms - sorry!

Reply:
The limit concurrency implementation I did recently acts more as a safeguard. The reason is that if you put 100 tasks in the queue and they hit the limit, they will start to retry and retry but only do so a couple of times and then fail. So this system isn't good for "slam stuff in the queue and wait for it to be done".

Setting up groups and chains allows us to control that there are 10 groups launched (= 10 running at the same time) and they sequentially check the alerts. That's a simple way I would say to control how much is done at a time.

Reply:
yep, super important, and generally better to stay within a tools expectations, just differing levels of celery skill and knowledge so we should make it super clear in the implementation (or safe in some other way tests/alerts/etc)

Reply:
About chains, a failure in the middle of a chain will make the whole chain fail, probably not what we want? So we either tweak tasks (e.g. catch all exceptions in the task?) or use another mechanism. 

About limit_concurrency, what if I modify it so that the tasks that don't meet the limit are retried using Celery retry? This solves the problem I described above and can be reused for other tasks.

Also, I looked a bit closer at Celery primitives, won't this all work using `chunks` only? Something like
`check_alert.chunks(all_alert_ids, 10)()`? IIUC, this will invoke 10 tasks, each with its own chunk of work, just like we want. There's a `group()` method for the chunks object, but I don't think it's necessary. The problem with the failures causing other failures remains.

And one more note, maybe the decision what to retry should not come from the task, because it's not synchronized. I'd say ideally it's the dependency's responsibility to decide what it can execute and it can't. But that's outside of this project.

So what do you say I use chunks + catch all exceptions and see how it works? Simple and should work.

Reply:
If we don't chain the chunks, we will have the effect of Celery picking up all chunks all at once, which is exactly what is to be prevented. Regarding exceptions: That's also useful. If there is some error e.g. regarding Clickhouse connection or something, we break the chain and all others don't need to try. Calculation errors (query syntax problems, timeouts) we can catch and store as error in the alert check.

Reply:
> If we don't chain the chunks, we will have the effect of Celery picking up all chunks all at once, which is exactly what is to be prevented.

I'm not sure this is true. I couldn't find anything in the documentation, so I looked in the implementation. `chunks` splits arguments into groups and then applies `xstarmap` to each group (see [this](https://github.com/celery/celery/blob/bd3b3c6afb8b0c44b416f469b3db6a725d71b609/celery/canvas.py#L1441-L1454)). `xstarmap` is essentially the `starmap` method (see [this](https://github.com/celery/celery/blob/bd3b3c6afb8b0c44b416f469b3db6a725d71b609/celery/app/task.py#L878)). And the starmap method runs sequentially as per the [documentation](https://docs.celeryq.dev/en/stable/userguide/canvas.html#map-starmap).

Also the chunk [documentation](https://docs.celeryq.dev/en/stable/userguide/canvas.html#chunks) hints that there will be only n tasks processing the load ("so that if you have one million objects, you can create 10 tasks with a hundred thousand objects each").

Maybe let's try using chunks alone and see how it goes? Say limit the number of concurrent executions to 1, add 10 alerts and see if they are executed sequentially? I was thinking about experimenting it locally, but I don't think it's easy to set up a fair experiment.

> Regarding exceptions: That's also useful. If there is some error e.g. regarding Clickhouse connection or something, we break the chain and all others don't need to try. Calculation errors (query syntax problems, timeouts) we can catch and store as error in the alert check.

I don't know about that, I would default to checking every alert. IMO it's bad to not execute an alert because it got in a wrong batch, it would hurt users. So I suggest here rethrow a couple types of exceptions that we're sure affect all alerts, but by default swallow exceptions, wdyt?

Reply:
> you can create 10 tasks

Then they will be running in parallel and might lead to high load.

>Say limit the number of concurrent executions to 1

I suggested a pretty straightforward way to start out and limit the load/concurrency for these alert checks. Since it wasn’t in your PR, I’ve already added it because it’s not complex and we can just try it. You seem to disagree and rule out doing it this way without offering an effective alternative to limit the load. Until it’s safe and won’t slam our cluster with dozens of calculations at the same time, I'm not sure how to proceed.

Reply:
> Then they will be running in parallel and might lead to high load.

I don't understand this, we want to detect anomalies in parallel as long as we control the number of tasks in flight, that's what I suggest. If you are claiming that "hundred thousand objects" will run in parallel, I don't think that's true, see my explanation about starmap above.

> What mechanism does this refer to?

Using chunks alone. `check_alert.chunks(all_alert_ids, len(all_alert_ids) / num_of_maximum_parallel_tasks)()`. IIUC the documentation and the implementation it will create `num_of_maximum_parallel_tasks` tasks, they will be executed in parallel, but every chunk will be executed sequentially, see what I wrote about starmap.

> I suggested a pretty straightforward

I'm worried about the straightforwardness of the approach. I need a couple minutes, a piece of paper and a pen to understand what it does, and I'm still not sure I understand it well.

My suggestion is simple - let's try chunks, run an experiment with num_of_maximum_parallel_tasks=1 and 10 alerts, and see if they are processed sequentially or in parallel.

Reply:
> they will be executed in parallel, but every chunk will be executed sequentially

That's right. And I get what you are suggesting now. However if only using chunks, imagine we have 1000 checks and chunk it by 10, that means the task will have a list of 100 insights to calculate. If this numbers grows and the average calculation time per insight is not super low, each task will run too long (thinking of deploys or other interruptions here). We could start out with that, I just imagine that we basically need to still improve this before releasing alerts more widely.

Reply:
Yeah, that's a valid point. Let's keep the chunks+chains+groups then.

Reply:
Ok, so about the interruptions you mentioned, I think chains have this problem too. Suppose a worker crashes while checking an alert (due to unrelated reasons), the whole chain will be cancelled, which is not good. We could mitigate it by using [acks_late](https://docs.celeryq.dev/en/stable/userguide/tasks.html#Task.acks_late), but then what if  the crash is caused by the alert check? Then retries will crash the worker indefinitely. I'm not sure either option is good.

 What if instead of splitting the load we limit the concurrency of the worker (like [this](https://github.com/PostHog/posthog/blob/2b1e1a5fb190bdd486229cec4a5bc89a3cd0b827/posthog/tasks/utils.py#L14))? IMO it's a much simpler approach, but I'm not sure it'll work since I don't know PostHog set up. Paul mentioned multiple pods, and it might become an issue, but maybe it's not that difficult to control and doesn't have many tradeoffs?

Reply:
You are right this would also break the chain 🤔 

I also thought the same if we can limit worker concurrency... the problem is, this concurrency setting is for one worker. We might spin up several pods though. And they also might autoscale to even more. Unfortunately we currently cannot control the worker amount and autoscaling granular per worker type.

I can think about it more, but maybe the case that alerts fail from time to time is not that severe for a first version?

Reply:
I suppose there's a way to create a dedicated pod with exact number of CPUs for this queue. But yeah, I see it's not that simple. Let's try the chains and see how it goes.

Reply:
It is indeed possible, but someone would need to dig into the configs and templates in the other repo 😓

Reply:
I just noticed that our infra team pushed an unrelated adjustment to auto scaling, which now allows for exact worker/pod configuration of maximum instances. I'll test this with some other tasks and we might end up just using the same mechanism here then after all 🎉

Reply:
Oh, cool!

What do you say we keep the mechanism with chains for now (to unblock development of other features and to merge that [old PR](https://github.com/PostHog/posthog/pull/22762/files)) and if the configuration works out I'll adjust it?